streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala

```diff
@@ -674,6 +676,8 @@ class StreamingContext private[streaming] (
           logWarning("StreamingContext has already been stopped")
         case ACTIVE =>
           scheduler.stop(stopGracefully)
+          // De-registering Streaming Metrics of the StreamingContext
+          env.metricsSystem.removeSource(streamingSource)
```

Comment:
The idea was to register at the call of the `start()`.  
So, based on your comment, that would mean registering the sources after the state is set to INITIALIZED and before.

`def start(): Unit = synchronized {
    // Registering Streaming Metrics at the start of the StreamingContext
    assert(env != null)
    assert(env.metricsSystem != null)
    env.metricsSystem.registerSource(streamingSource)`

Makes sense to have it after `INITIALIZED` and before the synchronized block of `ACTIVE` and `STOPPED`.
@tdas, could add more light.